addons/snailmail/models/snailmail_letter.py

```diff
@@ -343,18 +381,76 @@ def _snailmail_estimate(self):
+        for doc in docs:
+            if partner_field:
+                country_code = doc[partner_field].country_id.code
+            else:
+                country_code = doc.country_id.code
+            val = {
+            'pages': 1,
+            'address': {
+                'country_code': country_code,
+                },
+            }
+            doc_list.append(val)
+        params = {
+            'account_token': "",
+            'documents': doc_list,
+            'options': {
+                'color': self.env.user.company_id.snailmail_color,
+                'duplex': self.env.user.company_id.snailmail_duplex,
+                'currency_name': 'EUR',
+            },
+        }
+        req = jsonrpc(endpoint + '/iap/snailmail/1/estimate', params=params)
+        return req['total_cost']
+
     @api.model
     def _snailmail_cron(self):
-        letters_send = self.search([('state', '=', 'pending')])
-        if letters_send:
-            letters_send._snailmail_print()
-        limit_date = datetime.datetime.utcnow() - datetime.timedelta(days=1)
-        limit_date_str = datetime.datetime.strftime(limit_date, tools.DEFAULT_SERVER_DATETIME_FORMAT)
-        letters_canceled = self.search([
+        letters_send = self.search([
             '|',
-                ('state', '=', 'canceled'),
-                '&',
-                    ('state' ,'=', 'draft'),
-                    ('write_date', '<', limit_date_str),
+            ('state', '=', 'pending'),
+            '&',
+            ('state', '=', 'error'),
+            ('error_code', 'in', ['TRIAL_ERROR', 'CREDIT_ERROR', 'ATTACHMENT_ERROR', 'MISSING_REQUIRED_FIELDS'])
         ])
-        letters_canceled.unlink()
+        letters_send._snailmail_print()
+
+    @api.model
+    def fetch_failed_letters(self):
+        failed_letters = self.search([('state', '=', 'error'), ('user_id.id', '=', self.env.user.id), ('res_id', '!=', 0), ('model', '!=', False)])
+        return failed_letters._format_snailmail_failures()
+
+    @api.model
+    def _is_valid_address(self, record):
+        record.ensure_one()
+        required_keys = ['street', 'city', 'zip', 'country_id']
+        return all(record[key] for key in required_keys)
+
+    @api.multi
+    def _format_snailmail_failures(self):
+        """
+        A shorter message to notify a failure update
+        """
+        failures_infos = []
+        for letter in self:
+            info = {
+                'message_id': letter.message_id.id,
+                'record_name': letter.message_id.record_name,
+                'model_name': self.env['ir.model']._get(letter.model).display_name,
+                'uuid': letter.message_id.message_id,
+                'res_id': letter.res_id,
+                'model': letter.model,
+                'last_message_date': letter.message_id.date,
+                'module_icon': '/snailmail/static/img/snailmail_failure.png',
+                'snailmail_status': letter.message_id.snailmail_status,
```

Comment:
snailmail_error on message should be letter.state == 'error' no? 
same for status

Reply:
do you also want that we use letter.error_code if letter.state == 'error' for snailmail_status?

Reply:
If it works as expect yes.